zkevm-circuits/src/evm_circuit/execution/create.rs

```diff
@@ -0,0 +1,924 @@
+
+        let keccak_output = cb.query_word_rlc();
+        let create = ContractCreateGadget::construct(cb);
+        let contract_addr = expr_from_bytes(&keccak_output.cells[..N_BYTES_ACCOUNT_ADDRESS]);
+        let contract_addr_rlc = cb.word_rlc::<N_BYTES_ACCOUNT_ADDRESS>(
+            keccak_output
+                .cells
+                .iter()
+                .take(N_BYTES_ACCOUNT_ADDRESS)
+                .map(Expr::expr)
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        );
+
+        // stack operations
+        let value = cb.query_word_rlc();
+        let offset = cb.query_cell_phase2();
+        let length = cb.query_word_rlc();
+        cb.stack_pop(value.expr());
+        cb.stack_pop(offset.expr());
+        cb.stack_pop(length.expr());
+        cb.condition(is_create2.expr(), |cb| {
+            cb.stack_pop(create.salt_word_rlc(cb).expr());
+        });
+        cb.stack_push(is_success.expr() * contract_addr_rlc);
+
+        // read caller's balance and nonce
+        let caller_nonce = create.caller_nonce();
+        let caller_balance = cb.query_word_rlc();
+        cb.account_read(
+            create.caller_address(),
+            AccountFieldTag::Balance,
+            caller_balance.expr(),
+        );
+        cb.account_read(
+            create.caller_address(),
+            AccountFieldTag::Nonce,
+            caller_nonce.expr(),
+        );
+
+        // Pre-check: call depth, user's nonce and user's balance
+        let is_depth_in_range = LtGadget::construct(cb, depth.expr(), 1025.expr());
+        let is_insufficient_balance = LtWordGadget::construct(cb, &caller_balance, &value);
+        let is_nonce_in_range = LtGadget::construct(cb, caller_nonce.expr(), u64::MAX.expr());
+        let is_precheck_ok = and::expr([
+            is_depth_in_range.expr(),
+            not::expr(is_insufficient_balance.expr()),
+            is_nonce_in_range.expr(),
+        ]);
+
+        // verify gas cost
+        let init_code = MemoryAddressGadget::construct(cb, offset, length);
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [init_code.address()]);
+        let init_code_word_size = ConstantDivisionGadget::construct(
+            cb,
+            init_code.length() + (N_BYTES_WORD - 1).expr(),
+            N_BYTES_WORD as u64,
+        );
+        let keccak_gas_cost = init_code_word_size.quotient()
+            * select::expr(
+                is_create2.expr(),
+                (INIT_CODE_WORD_GAS + GasCost::COPY_SHA3).expr(),
+                INIT_CODE_WORD_GAS.expr(),
+            );
+        let gas_cost = GasCost::CREATE.expr() + memory_expansion.gas_cost() + keccak_gas_cost;
+        let gas_remaining = cb.curr.state.gas_left.expr() - gas_cost.clone();
+        let gas_left = ConstantDivisionGadget::construct(cb, gas_remaining.clone(), 64);
+        let callee_gas_left = gas_remaining - gas_left.quotient();
```

Comment:
Does there need to be also the [reduction](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L482) of gas (for `CreateDataGas`) for storing the code when there is no error?